### PR TITLE
Add e2e tests on bookmarking in a Collection page

### PR DESCRIPTION
**Pull request by gusaiani**
Adds tests in 

[frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js](https://github.com/metabase/metabase/pull/21058/files#diff-42959d74a9c718e9c6f57e785810dd2711904867527133251e8a00b32977dff4)
[frontend/test/metabase-visual/collections/bookmarks.cy.spec.js](https://github.com/metabase/metabase/pull/21058/files#diff-1271c12312aa01215413b4201915473cd811c963f4ed3b192682c984581ed3f8)

### CODE REVIEW

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (kulyk) [inline frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:25]: I think it'd make sense to cover this with unit tests using `color("brand")` instead of a hardcoded RGB value. This can make the test fragile

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (kulyk) [inline frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:53]: Same comment about color assertions as for the check above

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:63]: What's the difference between these two tests?

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:29]: Hard coding this could backfire in the future. To make it more resilient, let's import ADMIN from `cypress_data.js` and then construct this string using admin's first and last name, please.

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:1]: It's strange that linter doesn't complain, but I don't see `underscore` being used anywhere in this spec?

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:28]: We can (and should) rely more on a11y. Luckily, both "Bookmarks" and "Collections" are `H4`. I'd definitely utilize that for this test. It might seem fragile, but in this case it's good that it is. Because if we ever decide to NOT use HTML heading tags for this, test **should** fail and alert us, imo.
```js
function getSectionTitle(title) {
  return cy.findAllByRole("heading", { name: title });
}

// use it in test
sidebar().within(() => {
  getSectionTitle("Bookmarks");
  cy.findByText("Bobby Tables's Personal Collection");

  // Once there is a list of bookmarks,
  // we add a heading to the list of collections below the list of bookmarks
  getSectionTitle("Collections");
});
```

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:29]: Also, it might make sense to use `data-testid` for individual bookmarks in the sidebar (or even unordered list / table which would allow us to use `byRole` selectors).

**Comment by nemanjaglumac** [comment]
Just out of curiosity, are there plans to test removing bookmarks from the sidebar as well?
image

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (gusaiani) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:1]: Yes, strange miss by the linter. 

Removed.

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (gusaiani) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: Titles updated 👍

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (gusaiani) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: > or even unordered list / table

Agree that this would be better.

There's some necessary consistency with the collection links in the sidebar which may open Pandora's box, later with styling and so on.

Thinking about moving forward, are you ok with leaving the html tags as they are?

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (gusaiani) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: Yes, very strange. Removed.

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (gusaiani) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: 👍, updated

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (gusaiani) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: @kulyk makes 100% sense to me.

I went to unit tests and after quite some time, couldn't get styled colors from the svg.
All the basic ways of retrieving them failed.

So to be able to move forward, I've removed tests on colors.

Comment (kulyk) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: Thanks! For SVGs, you'd probably need to assert the `fill` property. On the other hand, I wonder if it'd be better to maybe add a visual Percy test for the whole collections page? We can catch a lof of visual regressions in this way: various bookmarks state, official collections, the collection contents layout, etc.

Comment (gusaiani) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: Yes @kulyk, I just added a visual regression test. Thank you.

File: frontend/test/metabase-visual/collections/bookmarks.cy.spec.js
Comment (kulyk) [inline frontend/test/metabase-visual/collections/bookmarks.cy.spec.js:21]: Please prefer using the API to set things up unless you're explicitly testing the UI flow. It makes the tests run much faster and it saves us a ton of CI time.

Also, since Percy visual tests cost some money, I think it'd make sense to create a single visual test for the whole collections page. We can do a few things:

* make one of the collection's official to also test the "badge" icon in the sidebar
* bookmark a few entities to cover more UI states (e.g. a question, a model, a dashboard, a collection, an official collection)
* add various entities to a collection that's going to be open during the snapshot and pin a few things to get some more UI states covered 

Everything listed above should be fairly easy to do via API. We can do that in a subsequent PR as well though.

Comment (gusaiani) [reply frontend/test/metabase-visual/collections/bookmarks.cy.spec.js]: @kulyk, thanks.

Bookmark now going via API.

Happy to add more checks in this test.
Soon I'll update the icons we render, and also the official badge, so it makes sense to add checks around those along with their implementation.

**Review by kulyk — APPROVED** [review]
LGTM! I'd wait for an approval from @nemanjaglumac though

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js:21]:
```suggestion
    cy.wait(["@fetchRootCollectionItems", "@fetchRootCollectionItems"]);
```

There are two such requests that we need to wait for. 
image

**Review by nemanjaglumac — APPROVED** [review]
🚀

File: frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/collections/bookmarks.cy.spec.js]: Yup. Leave them.

**Pull request merged**
